plugin/template: documentation - update the URL for caddyfile and add an example (#6213)
miekg/coredns: plugin/template/README.md

2. All queries will be answered (no `fallthrough`)
3. The answer is always NXDOMAIN

> ### Resolving SOA with dynamic HOSTNAME and static NS entries

Reviewer: ```suggestion
### Resolving SOA/NS for a dynamic domain
```

Return a start of authority record with a dynamic hostname:

~~~ corefile
engineering.coredns.io. {
    template IN SOA {
      match ^engineering\.coredns\.io\.$

>       answer "{{ .Zone }} 3600 IN SOA {$HOSTNAME}. engineering.coredns.io. (1 60 120 86400 3600)"

Reviewer: I don't think the SOA field values are correct here.

First field after record type should be the primary master namserver.  e.g. `ns1.engineering.coredns.io.`
Second field after record type should be an encoded admin email address. e.g. `admin.engineering.coredns.io.`

```suggestion
      answer "{{ .Zone }} 3600 IN SOA ns1.{{ .Zone }} admin.{{ .Zone }} (1 60 120 86400 3600)"
```

This would be the same for all returned SOA records.  But then, the dynamic hostname isn't used anywhere, so. I'm confused by the example. I'm probably missing something.

      fallthrough
    }
    template IN SOA {

>       authority "{{ .Zone }} 3600 IN SOA {$HOSTNAME}. engineering.coredns.io. (1 60 120 86400 3600)"

Reviewer: ```suggestion
      authority "{{ .Zone }} 3600 IN SOA ns1.{{ .Zone }} admin.{{ .Zone }} (1 60 120 86400 3600)"
```

    }
    template IN NS {
      match ^engineering\.coredns\.io\.$
      answer "{{ .Zone }} 900 IN NS ns1.engineering.coredns.io."
      answer "{{ .Zone }} 900 IN NS ns2.engineering.coredns.io."
      fallthrough
    }
    template IN NS {

>       authority "{{ .Zone }} 3600 IN SOA {$HOSTNAME}. engineering.coredns.io. (1 60 120 86400 3600)"

Reviewer: ```suggestion
      authority "{{ .Zone }} 3600 IN SOA ns1.{{ .Zone }} admin.{{ .Zone }} (1 60 120 86400 3600)"
```

    }
}
~~~

> 1. This template uses the defined zone (`engineering.coredns.io.`)

Reviewer: ```suggestion
1. This template uses the defined zone (`engineering.coredns.io.`)
2. The hostname used as the SOA record name is retrieved from the HOSTNAME environment variable. HOSTNAME contains a fully qualified domain name, in the zone `engineering.coredns.io.`.
```

> 2. All queries to SOA will be answered, the fallthrough on the first SOA block allows the second SOA block to be parsed as the match will only match subdomains and non-matching results will be handled by the second.

Reviewer: ```suggestion
3. All SOA/NS queries for the exact hostname will be answered with respective answers.
4. For subdomains of the hostname, the templates will `fallthrough` to the next template and return an empty answer (NODATA) and include the SOA record in the authority section (as required for negative answers).
```

> 3. A SOA record is sent with the local HOSTNAME defined by the environment.

Reviewer: ```suggestion
```

> 4. A pair of NS record is sent for any NS requests to the zone directly, while requests to subdomains will be served an SOA record.

Reviewer: ```suggestion
```
